apps/aens/src/aens_names.erl

```diff
@@ -0,0 +1,132 @@
+-export([owner/1,
+         status/1,
+         expires/1,
+         pointers/1,
+         ttl/1]).
+
+%%%===================================================================
+%%% Types
+%%%===================================================================
+-opaque name() :: #name{}.
+
+-type id() :: binary().
+-type serialized() :: binary().
+
+-export_type([id/0,
+              name/0,
+              serialized/0]).
+
+-define(NAME_TYPE, <<"name">>).
+-define(NAME_VSN, 1).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+-spec id(name()) -> aens_hash:name_hash().
+id(N) ->
+    hash(N).
+
+-spec new(aens_claim_tx:claim_tx(), non_neg_integer(), height()) -> name().
+new(ClaimTx, Expiration, BlockHeight) ->
+    Expires = BlockHeight + Expiration,
+    Hash = aens_hash:name_hash(aens_claim_tx:name(ClaimTx)),
+    %% TODO: add assertions on fields, similarily to what is done in aeo_oracles:new/2
+    #name{hash    = Hash,
+          owner   = aens_claim_tx:account(ClaimTx),
+          expires = Expires,
+          status  = claimed}.
+
+-spec update(aens_update_tx:update_tx(), name(), height()) -> name().
+update(UpdateTx, Name, BlockHeight) ->
+    Expires = BlockHeight + aens_update_tx:ttl(UpdateTx),
+    Name#name{expires  = Expires,
+              ttl      = aens_update_tx:name_ttl(UpdateTx),
+              pointers = aens_update_tx:pointers(UpdateTx)}.
+
+-spec revoke(name(), non_neg_integer(), height()) -> name().
+revoke(Name, Expiration, BlockHeight) ->
+    Expires = BlockHeight + Expiration,
+    Name#name{status  = revoked,
+              expires = Expires}.
+
+-spec transfer(aens_transfer_tx:transfer_tx(), name()) -> name().
+transfer(TransferTx, Name) ->
+    Name#name{owner = aens_transfer_tx:recipient_account(TransferTx)}.
+
+-spec serialize(name()) -> binary().
+serialize(#name{} = N) ->
+    msgpack:pack([#{<<"type">>     => ?NAME_TYPE},
+                  #{<<"vsn">>      => ?NAME_VSN},
+                  #{<<"hash">>     => hash(N)},
+                  #{<<"owner">>    => owner(N)},
+                  #{<<"expires">>  => expires(N)},
+                  #{<<"status">>   => status(N)},
+                  #{<<"ttl">>      => ttl(N)},
+                  #{<<"pointers">> => jsx:encode(pointers(N))}]).
+
+-spec deserialize(binary()) -> name().
+deserialize(Bin) ->
+    {ok, List} = msgpack:unpack(Bin),
+    [#{<<"type">>     := ?NAME_TYPE},
+     #{<<"vsn">>      := ?NAME_VSN},
+     #{<<"hash">>     := Hash},
+     #{<<"owner">>    := Owner},
+     #{<<"expires">>  := Expires},
+     #{<<"status">>   := Status},
+     #{<<"ttl">>      := TTL},
+     #{<<"pointers">> := Pointers}] = List,
+    #name{hash     = Hash,
+          owner    = Owner,
+          expires  = Expires,
+          status   = binary_to_existing_atom(Status, utf8),
+          ttl      = TTL,
+          pointers = jsx:decode(Pointers,[{labels, atom}])}.
```

Comment:
This looks like deserializing from state trees, and state trees in long chain may contain a lot of different labels. Could we avoid keeping them as atoms?

I assume labels are not in an enumeration. And if they are in an enumeration then we can use existing atoms only.

Reply:
The header file shows that currently we support only one atom. So, yes, it looks like we need to migrate do binary